File: windows/Marquee.Core/Models/BlocklistModels.cs
namespace Marquee.Core.Models;

// The request blocklist (api-v1.md section 7, "Request blocklist (0.41+)"):
// single titles and TMDb keywords/genres nobody may request. An older server
// leaves the title's viewer.blocked out and answers 404 for
// /settings/blocklist; the app then shows none of this.

/// <summary>
/// <c>viewer.blocked</c> (0.41+): the title is on the admin's blocklist, so
/// <c>canRequest</c>, <c>canRequestSeasons</c> and <c>fourK.canRequest</c>
/// are already false. <see cref="Keyword"/> is set when a keyword or genre
/// did it (then it can only be unblocked from Settings).
/// </summary>
public sealed record TitleBlock
{
    /// <summary>The admin's reason, shown to whoever asks; null when none was given.</summary>
    public string? Reason { get; init; }

    /// <summary>The TMDb keyword or genre that blocked it, lower-case; null when the title itself is blocked.</summary>
    public string? Keyword { get; init; }

    /// <summary>A member's pill (components/add-to-library-button.tsx): "Requests are closed for this title — Already on Max."</summary>
    public string MemberLine =>
        Reason.NonBlank() is { } reason ? $"Requests are closed for this title — {reason}" : "Requests are closed for this title";

    /// <summary>The admin's pill when a keyword did it: "Requests blocked by “anime”"; null for a title blocked directly.</summary>
    public string? KeywordLine => Keyword.NonBlank() is { } keyword ? $"Requests blocked by “{keyword}”" : null;
}

public readonly record struct BlocklistKind(string Value) : IOpenEnum<BlocklistKind>
{
    public static readonly BlocklistKind Title = new("title");
    public static readonly BlocklistKind Keyword = new("keyword");

    public static IReadOnlyList<BlocklistKind> Known { get; } = [Title, Keyword];
    public static BlocklistKind FromValue(string value) => new(value);
    public bool IsKnown => Known.Contains(this);
    public override string ToString() => Value;
}

/// <summary>One row of <c>GET /settings/blocklist</c> (admin): keywords first, then titles.</summary>
public sealed record BlocklistEntry
{
    public required Guid Id { get; init; }
    public required BlocklistKind Kind { get; init; }

    /// <summary>A title's type and id; null for a keyword.</summary>
    public MediaType? MediaType { get; init; }
    public int? TmdbId { get; init; }

    /// <summary>A title's name as it was when blocked.</summary>
    public string? Title { get; init; }

    /// <summary>A TMDb keyword or genre, lower-case.</summary>
    public string? Keyword { get; init; }

    public string? Reason { get; init; }
    public required DateTimeOffset CreatedAt { get; init; }

    /// <summary>The title this row links to; null for a keyword (or a title row missing its id).</summary>
    public TitleId? TitleId =>
        Kind == BlocklistKind.Title && MediaType is { } type && TmdbId is { } id ? new TitleId(type, id) : null;

    /// <summary>What the website's list prints: the title's name (or "#438631"), else "Keyword: anime".</summary>
    public string Label
    {
        get
        {
            if (TitleId is { } title)
            {
                return Title.NonBlank() ?? $"#{title.TmdbId}";
            }
            return $"Keyword: {Keyword}";
        }
    }
}

/// <summary><c>POST /titles/{type}/{tmdbId}/block</c> body; without a reason no body is sent.</summary>
public sealed record BlockTitleBody(string Reason);

/// <summary><c>POST /settings/blocklist</c> body; a blank reason is left out.</summary>
public sealed record BlockKeywordBody(string Keyword, string? Reason = null);

File: mac/Marquee/API/Models/BlocklistModels.swift
import Foundation

// The request blocklist (api-v1.md §7 "Request blocklist (0.41+)"): titles and
// TMDb keywords/genres nobody may request. A server older than 0.41 has none
// of it (`viewer.blocked` is missing, `GET /settings/blocklist` is a 404), and
// the app then shows none of it.

extension API {
    /// `viewer.blocked`: this title is on the admin's blocklist.
    struct TitleBlock: Codable, Hashable, Sendable {
        /// The admin's reason, shown to whoever asks.
        let reason: String?
        /// Set when a blocked keyword or genre did it (unblock it in Settings).
        let keyword: String?

        /// The member's pill: "Requests are closed for this title — Already on Max."
        var closedLine: String {
            if let reason = reason.nonBlank {
                return "Requests are closed for this title — \(reason)"
            }
            return "Requests are closed for this title"
        }
    }

    /// `viewer.blocked` as sent: `null` is `.notBlocked`, so a server that
    /// omits the key (older than 0.41, which can't block) stays tellable
    /// apart as a nil `TitleViewerState.blocked`.
    enum BlockState: Codable, Hashable, Sendable {
        case notBlocked
        case blocked(TitleBlock)

        var block: TitleBlock? {
            if case let .blocked(block) = self { return block }
            return nil
        }

        init(from decoder: Decoder) throws {
            let container = try decoder.singleValueContainer()
            self = container.decodeNil() ? .notBlocked : .blocked(try container.decode(TitleBlock.self))
        }

        func encode(to encoder: Encoder) throws {
            var container = encoder.singleValueContainer()
            switch self {
            case .notBlocked: try container.encodeNil()
            case let .blocked(block): try container.encode(block)
            }
        }
    }

    enum BlocklistKind: OpenEnum {
        case title
        case keyword
        case unknown(String)

        static let knownCases: [BlocklistKind] = [.title, .keyword]

        var rawValue: String {
            switch self {
            case .title: return "title"
            case .keyword: return "keyword"
            case let .unknown(raw): return raw
            }
        }
    }

    /// One row of `GET /settings/blocklist` (keywords first, then titles).
    struct BlocklistEntry: Codable, Hashable, Sendable, Identifiable {
        /// Kept as the server's string, like `IssueReporter.userId`.
        let id: String
        let kind: BlocklistKind
        /// Titles only.
        let mediaType: MediaType?
        let tmdbId: Int?
        /// The title's name when it was blocked.
        let title: String?
        /// Keywords only, lower-case.
        let keyword: String?
        let reason: String?
        let createdAt: Date

        /// The title page this row links to; nil for a keyword.
        var titleID: TitleID? {
            guard kind == .title, let mediaType, let tmdbId else { return nil }
            return TitleID(mediaType, tmdbId)
        }

        /// The title's name ("#438631" without one), or "Keyword: anime".
        var label: String {
            if titleID != nil, let tmdbId {
                return title.nonBlank ?? "#\(tmdbId)"
            }
            return "Keyword: \(keyword ?? "")"
        }
    }

    /// `POST /titles/{type}/{tmdbId}/block` body. nil `reason` sends no key.
    struct BlockTitleRequest: Codable, Hashable, Sendable {
        let reason: String?

        /// The server's limit on `reason`.
        static let maxReasonLength = 200
    }

    /// `POST /settings/blocklist` body.
    struct BlockKeywordRequest: Codable, Hashable, Sendable {
        let keyword: String
        let reason: String?
    }
}

extension KeyedDecodingContainer {
    /// `viewer.blocked`: a present `null` is `.notBlocked` rather than nil
    /// (the synthesized `TitleViewerState` decoder picks this overload).
    func decodeIfPresent(_ type: API.BlockState.Type, forKey key: Key) throws -> API.BlockState? {
        guard contains(key) else { return nil }
        if try decodeNil(forKey: key) { return .notBlocked }
        return .blocked(try decode(API.TitleBlock.self, forKey: key))
    }
}
